### DISCUSSION TITLE
Vlasov equation integration issue

### DISCUSSION THREAD

**Discussion by yl3dy** [hyperdeal/hyperdeal#90] [General]
I'm currently trying to develop a 6-D Vlasov FEM solver using hyper.deal accounting for both electric and magnetic field as a first step towards a self-consistent Vlasov-Maxwell solver. The current physical problem is collisionless plasma in a constant magnetic field B. From my understanding, spherically symmetric velocity distribution should not change in time in this case (magnetic field would induce its rotation around B vector).

Diagnostics tools are density ($\int f(r,v) dv$), velocity distribution ($\int f(r,v) dr$), full f(r,v) output (r, v and f(r,v) at each quadrature point) and advection vector (transport velocity) field. The issue is that when using a function depending on v as an advection vector in the df/dv term of the Vlasov equation artifacts appear (below the examples are for the Lorenz force without electric field, but it is observable for something artificial like $(v \cdot B) B$ too). They manifest themselves as ever increasing values "along" magnetic field for odd quadrature points in velocity distribution plots (see attached images). What the strangest thing is that saving solution values at quadrature points, doing naïve integration (i.e. just summing f values) and plotting using Gnuplot does _not_ reveal such artifacts: the velocity distribution remains spherically symmetric. It is puzzling that their amplitude does _not_ depend on the usual numerical parameters: dt and grid density.

See `artifact_vtu.png` for velocity distribution calculated using deal.II/hyper.deal routines, `artifact_naive.png` for a naïve velocity distribution calculation. Advection vector (transport velocity) fields for coordinate and velocity spaces are shown in `advection_x.png` at v = (0.014088, -0.38909, -0.36091) and in `advection_v.png` at r = (-0.014088, -0.014088, -0.0625). Numerical parameters are dt = 0.01, T = 0.1, the grid is a hypercube [-0.5; 0.5]^6 with the size of 8 subdivisions along each dimension. Magnetic field is constant B = (1, 0, 0), normalized in such a way that $F_{Lorenz} = q [v \times B]$.

The velocity distribution routine is implemented along the same lines as `hyperdeal::VectorTools::velocity_space_integration` (`include/vector_tools.h`):

```C++
data.template cell_loop<Vector_Out, Vector_In>(
  [&](const auto&, Vector_Out& dest, const Vector_In& src_,
      const auto cell) mutable
  {
    phi.reinit(cell);
    phi.read_dof_values(src_);
    phi_x.reinit(cell.x);
    phi_v.reinit(cell.v);

    const VectorizedArrayType* data_ptr_src = phi.get_data_ptr();
    VectorizedArrayTypeV* data_ptr_dst = phi_v.begin_dof_values();

    const auto nv_factor = dealii::Utilities::pow(n_points, dim_v);
    const auto N_lanes = phi.n_vectorization_lanes_filled();

    for (unsigned int qv = 0; qv < phi_v.n_q_points; ++qv) {
      VectorizedArrayTypeV sum_x = 0.0;

      for (unsigned int qx = 0; qx < phi_x.n_q_points; ++qx) {
        const auto q = qx + qv*nv_factor;
        for (unsigned int lane = 0; lane < N_lanes; ++lane)
          sum_x += data_ptr_src[q][lane] * phi_x.JxW(qx)[lane];
      }

      data_ptr_dst[qv] = sum_x;
    }

    phi_v.distribute_local_to_global(dest);
  },
  dst,
  src
);
```

The extraction of solution values at quadrature points is quite similar (`include/output.h`, line 427):

```C++
matrix_free.template cell_loop<int, VectorType>(
  [&out, &phi, &phi_x, &phi_v](const auto&, int&,
                               const VectorType& src,
                               const auto cell) mutable
  {
    phi.reinit(cell);
    phi.read_dof_values(src);
    phi_x.reinit(cell.x);
    const unsigned int index_v = cell.v / VectorizedArrayTypeV::size();
    const unsigned int lane_v = cell.v % VectorizedArrayTypeV::size();
    phi_v.reinit(cell.v);

    const VectorizedArrayType* data_ptr_src = phi.get_data_ptr();

    const auto nv_factor = dealii::Utilities::pow(n_points, dim_v);
    const auto N_lanes = phi.n_vectorization_lanes_filled();
    for (unsigned int qv = 0; qv < phi_v.n_q_points; ++qv) {
      for (unsigned int qx = 0; qx < phi_x.n_q_points; ++qx) {
        const auto qx_point = phi_x.quadrature_point(qx);
        const auto qv_point = phi_v.quadrature_point(qv);
        const auto q_value = data_ptr_src[qx + qv*nv_factor];

        for (unsigned int lane = 0; lane < N_lanes; ++lane) {
          double cache = 0.0;
          // Write coordinates
          for (unsigned int d = 0; d < dim_x; ++d) {
            // all lanes filled
            cache = qx_point[d][lane];
            out.write(reinterpret_cast<const char*>(&cache), sizeof cache);
          }
          for (unsigned int d = 0; d < dim_v; ++d) {
            // only one lane used for v
            cache = qv_point[d][lane_v];
            out.write(reinterpret_cast<const char*>(&cache), sizeof cache);
          }
          // Write value
          cache = q_value[lane];
          out.write(reinterpret_cast<const char*>(&cache), sizeof cache);
        }
      }
    }
  },
  dummy, solution
);
```

Attached below is also a tarball `plasma-model.tar.bz2` with the complete buildable C++17 sources. The plots can be recreated by running `vlasov` binary with `--output-full` argument via `mpirun` (*the output is about 36 GiB large!*). VTUs can be viewed using `vlasov_state.pvsm` in ParaView, naïve integration by running `reader -i 10 --integrate` and then plotting `density.gpi` with Gnuplot.

What advice can you give on debugging the issue? Are there flaws in my understanding of deal.II/hyper.deal or numerical peculiarities of such a model?

Best regards,
Alexander Kiselyov

P.S. System specifications:

* A single machine with Intel Core i7-12700K CPU
* NixOS 21.11 amd64 (Linux kernel 5.16.3, gcc 10.3.0, OpenMPI 4.1.1)
* Deal.II 10.0.0-prerelease (b01ae8a)
* Hyper.deal (9bfd4b2)
* ParaView 5.8.0

![artifact_vtu](https://user-images.githubusercontent.com/1311192/152348673-3ae72ee6-b862-498a-ab35-37d61c482e4b.png)
![artifact_naive](https://user-images.githubusercontent.com/1311192/152348692-40bd539d-f2e0-42bc-8cd7-6a0098fa63bd.png)
![advection_x](https://user-images.githubusercontent.com/1311192/152348700-fed0fb3d-92ae-4d6b-8b20-248be7fe2044.png)
![advection_v](https://user-images.githubusercontent.com/1311192/152348713-62d4c6e9-070f-4de3-8bbd-63185f98da36.png)
[plasma-model.tar.gz](https://github.com/hyperdeal/hyperdeal/files/7994795/plasma-model.tar.gz)

**Comment by peterrum** [comment]
Hi Alexander,

my suggestion would that you write a test similar to https://github.com/hyperdeal/hyperdeal/blob/master/tests/vector_tools/vector_tools_01.cc for your function.
I would hope that we can reproduce the issue in a simplified context and in lower dimensional space!

Thanks,
Peter

**Reply by yl3dy** [reply]
I'm not really sure how to apply that kind of test here. The thing that puzzles me most is that doing integration with these two methods (via `dealii::FEEvaluation` and just writing quadrature values to file and manually summing them) gives starkly different results for the same solution vector. As for a simplified context, should I just save the DoF vector of a solution (e.g. the one mentioned in the OP) and run through both integration methods? Should I explicitly compare values at quads 0 (Gauss-Legendre) and 2 (Gauss-Lobatto)?

**Reply by peterrum** [reply]
The idea of the test is that we do not have to perform a complete simulation but can test the new function on its own. I would suggest that I write a test for `VectorTools::velocity_space_integration()` (which is apparently missing) which you could simply extend for you function! I have today a tight schedule but hope that I have time tomorrow! Does that work for you?

> As for a simplified context, should I just save the DoF vector of a solution (e.g. the one mentioned in the OP) and run through both integration methods? Should I explicitly compare values at quads 0 (Gauss-Legendre) and 2 (Gauss-Lobatto)?

The reason why we are using `2 (Gauss-Lobatto)` here is that the DoFs are positioned at these points and as a consequence we can skip the interpolation here. In other places of the library (advection operator), the interpolation is optional.

By the way, would you be interested to make your function part of `hyper.deal`?

**Reply by yl3dy** [reply]
I've already started coding a (hopefully) simple test loading a saved solution vector from the main simulation code and calculating velocity distribution by two methods, but still have to iron out some coding issues. I suppose it's more or less what you mean.

> I would suggest that I write a test for VectorTools::velocity_space_integration() (which is apparently missing) which you could simply extend for you function!

That would be great!

> The reason why we are using 2 (Gauss-Lobatto) here is that the DoFs are positioned at these points and as a consequence we can skip the interpolation here.

Actually I wanted to ask a question about different quadratures too! I was a bit confused by DoF notion of deal.II in the matrix-free case. It seems that `DoFHandler` doesn't take quadrature formulas into account, but `MatrixFree` does support multiple quadratures with (possibly) different points with the same `DoFHandler`. So how exactly DoF vector is interpreted in this case?

> By the way, would you be interested to make your function part of hyper.deal?

I'd be very happy to submit a pull request after debugging is done!

**Reply by kronbichler** [reply]
> Actually I wanted to ask a question about different quadratures too! I was a bit confused by DoF notion of deal.II in the matrix-free case. It seems that `DoFHandler` doesn't take quadrature formulas into account, but `MatrixFree` does support multiple quadratures with (possibly) different points with the same `DoFHandler`. So how exactly DoF vector is interpreted in this case?

`MatrixFree` or actually the evaluator on top of it, `FEEvaluation`, provides you with an interpretation of the solution at a specific set of *points*, whereas `DoFHandler` provides a mere numbering of the degrees of freedom. So when you ask for one or the other quadrature formula (Gauss or Gauss-Lobatto), you compute interpolations of the solutions (described by the pair of a vector and `DoFHandler`) at different set up points, and it is up to the application to decide what is useful. As @peterrum said above, the Gauss-Lobatto case has the special property of having the quadrature points aligned with the points where we define Lagrange polynomials, so the process of getting the solution values at these points is simply a copy from the global vector entries to the local quadrature points, whereas the Gauss case also involves an interpolation step. Note that in deal.II we provide flexibility with respect to the points of the quadrature formula (where `FEEvaluation` will give you the data) and the points used for the polynomial representation. This is opposed to some other spectral element codes that hardcode the position of the points of evaluation to the support points of the Lagrange polynomials.

**Reply by yl3dy** [reply]
@kronbichler Thanks a lot for the explanation!

@peterrum I've hacked together a [short test](https://gist.github.com/yl3dy/680e0c0e8d9d21827453e4b5cd224c08). It reads a binary dump of a solution vector and builds velocity distribution using two methods outlined above. The blobs can be downloaded [here](https://cloud.kiselyov.cc/s/3Je5e3dHd8rXLQG), they're `vct_solution_??_??.bin` files. These were generated using the default settings in the code attached in the original message.

**Reply by peterrum** [reply]
> That would be great!

@yl3dy  I have added the promised test(s) in https://github.com/hyperdeal/hyperdeal/pull/92. For the time being, I have also added `VT::coordinate_space_integration()` (slightly modified version of your code - however I plan to remove it so that you can create a PR). The second test program also outputs the data with `DataOut`. Everything looks to me fine. Could you check out the code? I have the impression that your function is fine and something else goes wrong (so that the solution has some discontinuities!?).

Do you the same issue when you output the initial condition?

**Reply by yl3dy** [reply]
@peterrum Thank you very much for the test! These tests pass here too, which is certainly encouraging.

But the original question was also about a regular VTK-based output versus saving solution values at quadrature points as a crosscheck. As far as I understand, plotting such values on a heatmap should give a quite close approximation of a VTK output given that the element degree is 2. In practice, the heatmap approach gives a "correct" plot (i.e. the Maxwellian "fuzzy ball"), while sharp extrema are visible in VTK. This makes me wonder what is the reason for this discrepancy.

The script I linked earlier does exactly that over a saved DoF vector: generates both VTK (`hyperdeal_like_integration`) and per-quadrature-point outputs (`naive_integration`). The thing I want to be sure of is that these sharp maxima are a real feature of the solution, not a plotting artifact.

A probable piece of evidence might be the way I build patches. I was using this piece of code basically copy-pasted from hyper.deal examples:

```C++
dealii::DataOutBase::VtkFlags flags;
flags.write_higher_order_cells = true;
data_out.set_flags(flags);
// ...
data_out.build_patches(dealii::MappingQGeneric<dim_v>(3), 3,
                       dealii::DataOut<dim_v>::CurvedCellRegion::curved_inner_cells);
```
Curiously, if building patches with default settings (no higher order cells and no custom Mapping), the picture shows actual discontinuities:
![artifact_no_higher_order_cells](https://user-images.githubusercontent.com/1311192/152842170-03a6ea11-9f25-4900-af49-8ee960040800.png)

> Do you the same issue when you output the initial condition?

No, the issue is absent then.

**Comment by kkormann** [comment]
Hi Alexander,
I am currently developing a Vlasov-Maxwell solver within hyper.deal. Maybe we can have a discussion together? I contacted you by e-mail.
Best regards,
Katharina

**Comment by yl3dy** [comment]
A minor question about the same issue: I want to clarify a point about correctly building patches for graphical output.

In the current model I'm using a hypercubic mesh with rectilinear cells, consequently the mapping in `MatrixFree`s is `MappingQGeneric<dim_x>(1)`. To output the data, previously I used the same sequence as in hyperdeal examples: enabled `write_higher_order_cells` in `VtkFlags` and used `data_out.build_patches(MappingQGeneric<dim_x>(3), 3);`. As far as I understand, specifiying higher order mapping should give the same results if its degree is equal or higher than the one in the `MatrixFree`'s mapping, and adding subdivisions also gives the same results for non-curved cells. This procedure gives the density image as in the original post. It seems that additional elements are fitted for each subdivision, which might make sharp tears look continuous.

When disabling higher order cells and using 1 subdivision, the picture is starkly different (see my last comment in the first thread): element discontinuity is apparent. On the other hand, even when plotting the initial condition of the symmetric function (here it is `pow(20 / PI, 1.5) * exp(-20*x.norm_squared())`) elements are apparently aligned with `y = x` diagonal. See the attached image, gridlines are cell borders (here I used `write_higher_order_cells = false` and `data_out.build_patches(MappingQGeneric<dim_x>(1), 1)`).

@peterrum Is my understanding of the output process lacking? Which of these methods is more accurate to the actual solution?

![initial_with_1_subdivision](https://user-images.githubusercontent.com/1311192/153475711-e2200f82-df15-46d1-92e1-bd088e0d761c.png)

**Reply by kronbichler** [reply]
The figure here with the `x=y` is mostly an artifact of the visualization program. ParaView plots bilinear functions by splitting a quadrilateral into two triangles and then plotting the solution on those.

To answer some questions:
> As far as I understand, specifiying higher order mapping should give the same results if its degree is equal or higher than the one in the MatrixFree's mapping,

For Cartesian meshes, the degree of the mapping should not matter at all, both in the computation and in the visualization `MappingQGeneric(1)` and `MappingQGeneric(3)` should be equivalent. This is because higher order mappings simply add additional points, and in the absence of additional information those just get aligned with the cells. Things get different once manifolds are attached to the `Triangulation` (deal.II does that for most of the grids produced by `GridGenerator`, and you can also do that manually by `add_manifold` as done e.g. https://github.com/hyperdeal/hyperdeal/blob/ab48f28b49fe4353153bdffbbaf27216203a3fe7/examples/advection/cases/torus_hyperball.h#L100-L110).

So I do not think that the mapping should make a difference in your case, unless there is some manifold hidden somewhere.

Regarding the artifacts in the other post, I believe this must be some feature in the solution. It is strange that you can't see the same trend, sampled on more points, in the higher order output, as there shouldn't be a difference in how the solution gets interpreted by ParaView as far as I know.

**Reply by yl3dy** [reply]
Thank you very much for the answer!

I've already coded a simplified interpolation example using only Deal.II, which gives the same result. Seems like it really is ParaView's artifacts: calculating function values at a finer grid using `dealii::VectorTools::point_value` and plotting it as a raster image provides a much smoother picture. Moreover, it changes in an expected manner when the element degree is changed, while the plot in ParaView/Mayavi stays the same regardless of it. By the way, when plotting the same VTU in Mayavi (version 4.7.3 here) the picture is almost the same (it's exactly the same when enabling the option "Interpolate scalars before mapping" in "Actor → Mapper" in "Surface" module, although even without it the triangular artifacts are clearly visible).

What visualization software could you recommend which is friendlier towards Deal.II output?

**Reply by kronbichler** [reply]
I do use `ParaView` mostly, so I have mostly learned to live with the defects in the way it represents things. I like the high-order output because it allows to refine the "mesh" used for visualization (for higher order functions).
